frontend/src/components/Common/FormModal.tsx

```diff
@@ -0,0 +1,157 @@
+import {
+  Button,
+  type ButtonProps,
+  FormControl,
+  FormErrorMessage,
+  FormLabel,
+  Input,
+  Modal,
+  ModalBody,
+  ModalCloseButton,
+  ModalContent,
+  ModalFooter,
+  ModalHeader,
+  ModalOverlay,
+  Textarea,
+} from "@chakra-ui/react"
+import { useMutation, useQueryClient } from "@tanstack/react-query"
+import { type SubmitHandler, useForm } from "react-hook-form"
+
+import type { ApiError } from "../../client"
+import useCustomToast from "../../hooks/useCustomToast"
+import { handleError } from "../../utils"
+
+interface FieldDefinition<T> {
+  name: keyof T
+  label: string
+  placeholder?: string
+  type?: string
+  validation?: any
+  defaultValue?: any
+  inputProps?: any
+}
+
+interface ExtendedButtonProps extends ButtonProps {
+  "data-testid"?: string
+}
+
+export interface FormModalProps<T> {
+  isOpen: boolean
+  onClose: () => void
+  title: string
+  fields: FieldDefinition<T>[]
+  mutationFn: (data: T) => Promise<any>
+  successMessage: string
+  queryKeyToInvalidate?: string[]
```

Comment:
nit: change this variable name to `queryKey`

Reply:
It indicates which data (identified by query keys for each data fetch request) should be fetched again from the server after a successful change, ensuring the information remains current.

Reply:
It's a minor point.  The string "item" is the string Tanastack Query uses to keep track of transactions.  What do with it in the modal is InvalidateIt.  But the QueryString can be used for other purposes -- it does not exist only for invalidation.

This is an academic argument, though.  Either way is fine.